### PR TITLE
Add dataframe performance doc

### PR DESCRIPTION
**Pull request by mrocklin**
cc @TomAugspurger @jcrist @teoliphant @sinhrks @martindurant 

Any other recommendations or changes here?  Direct edits to this PR are welcome.

### CODE REVIEW

File: docs/source/dataframe-performance.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+Repartition to Reduce Overhead
+------------------------------
+
+Your Dask.dataframe is split up into many Pandas dataframes.  We sometimes call
+these "partitions".  Often the number of partitions is decided for you; for
+example it might be the number of CSV files from which you are reading. However
+over time as you reduce or increase the size of your pandas dataframes by
+filtering or joining it may be wise to reconsider how many partitions you need.
+There is a cost to having too man or having too few.
```
Comment (TomAugspurger) [inline docs/source/dataframe-performance.rst]: typo: man -> many

File: docs/source/dataframe-performance.rst [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+Partitions should fit comfortably in memory (smaller than a gigabyte) but also
+not be too numerous.  Every operation on every partition takes the central
+scheduler a few hundred microseconds to process.  If you have a few thousand
+tasks this is barely noticeable, but it is nice to reduce the number is
```
Comment (TomAugspurger) [inline docs/source/dataframe-performance.rst]: typo: is -> if

File: docs/source/dataframe-performance.rst [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+
+Joins
+-----
+
+Joining two dataframes can be either very expensive or very cheap depending on
+the situation.  It is cheap in the following cases:
+
+1.  Joining a Dask.dataframe with a Pandas dataframe
+2.  Joining a Dask.dataframe with a Dask.datafrmae of a single partition.
```
Comment (TomAugspurger) [inline docs/source/dataframe-performance.rst]: typo: Dask.datafrmae -> Dask.dataframe

**Comment by mrocklin** [comment]
Thanks for the comments and proof reading @TomAugspurger (sorry for the typos!)

One question about this:

> If you happen to know the set of categories ahead of time, but are reading from a source that doesn't store the categorical metadata (CSV), you can use df.col.cat.set_categories(categories) to avoid a full scan of the data

Given @jcrist 's recent changes allowing different categoricals between partitions is this still a common performance issue?

**Comment by TomAugspurger** [comment]
I supposes it's less common now, but should still matter in cases like get_dummies. This raises on unknown categories. I think a .set_caregories(cats) would be cheaper than a .as_known, but I haven't tested. 




On Sat, Feb 4, 2017 at 2:49 PM -0600, "Matthew Rocklin" <notifications@github.com> wrote:












Thanks for the comments and proof reading @TomAugspurger (sorry for the typos!)


One question about this:



If you happen to know the set of categories ahead of time, but are reading from a source that doesn't store the categorical metadata (CSV), you can use df.col.cat.set_categories(categories) to avoid a full scan of the data



Given @jcrist 's recent changes allowing different categoricals between partitions is this still a common performance issue?

**Comment by jcrist** [comment]
> I think a .set_caregories(cats) would be cheaper than a .as_known, but I haven't tested.

`set_categories(cats)` requires no computation at graph build time, and so will be *significantly* cheaper than `as_known()` or `categorize()`. It does require knowing the categories you want beforehand though.

**Comment by mrocklin** [comment]
I've fixed typos.  I would like to delay the set_categories bit, mostly out of laziness.  It seems important, but not common-case enough to be essential.  Is this OK?

**Comment by jcrist** [comment]
LGTM. `set_categories` has the same semantics as in pandas, so I think adding a special note for dask is unnecessary at this time.

**Pull request merged**